### ISSUE TITLE
Ｉ can't write data to win-tap

### ISSUE THREAD

**Issue by finddiff** [songgao/water#40]
here is my code:

package main

import (
	"log"
	"github.com/songgao/packets/ethernet"
	"github.com/songgao/water"
	"fmt"
)

func main() {
	ifce, err := water.New(water.Config{
		DeviceType: water.TAP,
		PlatformSpecificParams:water.PlatformSpecificParams{
			ComponentID: "tap0901",
			Network:     "192.168.1.10/24",
		},
	})
	if err != nil {
		log.Fatal(err)
	}
	var frame ethernet.Frame

	for {
		frame.Resize(1500)
		n, err := ifce.Read([]byte(frame))
		if err != nil {
			log.Fatal(err)
		}
		frame = frame[:n]
		fmt.Printf("\nDst: %s\n", frame.Destination())
		fmt.Printf("Src: %s\n", frame.Source())
		fmt.Printf("Ethertype: % x\n", frame.Ethertype())
		n, err =ifce.Write(frame)// it block here ,not returen anythin
		if err != nil {
			log.Fatal(err)
		}
	}
}

**Comment by lixin9311** [comment]
I'll look into this later

**Comment by lixin9311** [comment]
```go
package main

import (
	"log"

	"github.com/songgao/packets/ethernet"
	"github.com/songgao/water"
)

func main() {
	ifce, err := water.New(water.Config{
		DeviceType: water.TAP,
	})
	if err != nil {
		log.Fatal(err)
	}
	var frame ethernet.Frame

	for {
		frame.Resize(1500)
		n, err := ifce.Read([]byte(frame))
		if err != nil {
			log.Fatal(err)
		}
		frame = frame[:n]
		log.Printf("Dst: %s\n", frame.Destination())
		log.Printf("Src: %s\n", frame.Source())
		log.Printf("Ethertype: % x\n", frame.Ethertype())
		log.Printf("Payload: % x\n", frame.Payload())
		if _, err := ifce.Write(frame); err != nil {
			log.Fatal(err)
		} else {
			log.Println("Frame written")
		}
	}
}
```
This example code works well on my Windows 10. And I even tried a simple VPN code, everything just works fine.

**Comment by songgao** [comment]
Closing due to inactivity.

**Comment by wwwzbwcom** [comment]
Same here, program dead after write to interface, really strange

**Comment by gordonguilloryklx** [comment]
This is a confusing one. I've tried everything but cannot get the tap interface to receive my writes.

![image](https://github.com/songgao/water/assets/167548505/369a9b3a-b6b8-4842-bb02-03bcf5af4fae)

**Comment by songgao** [comment]
> This is a confusing one. I've tried everything but cannot get the tap interface to receive my writes.
> 
> ![image](https://private-user-images.githubusercontent.com/167548505/327266075-369a9b3a-b6b8-4842-bb02-03bcf5af4fae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ2MDUzMzQsIm5iZiI6MTcxNDYwNTAzNCwicGF0aCI6Ii8xNjc1NDg1MDUvMzI3MjY2MDc1LTM2OWE5YjNhLWI2YjgtNDg0Mi1iYjAyLTAzYmNmNWFmNGZhZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUwMVQyMzEwMzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNzI5NDYzMmQxZDg5NzUzN2ZhYjA4NzAzNmE3MmI0Y2UzOGQwZmUwNDE5YmRhYjYwYjNiMTA3NzljOTE1MTdmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.l4NM9FtcPZTOB5MMKfJStw-ePNcNm8zyjfyy5ZdNT_4)

@gordonguilloryklx try flipping src and dst before writing it back? I wonder if it’s the kernel that simply dropped it because it the addresses were wrong

**Comment by zveinn** [comment]
Remember to flip both IP and port, also.. you might need to re-calculate the header checksums again. 
gopacket is good for that.

**Comment by gordonguilloryklx** [comment]
I did that here, but I am just seeing @zveinn 's message and wondering if that's why it is showing 0 received and many sent now.

![image](https://github.com/songgao/water/assets/167548505/4b993770-3146-4a3d-8f82-820d2ad01c51)

**Comment by zveinn** [comment]
seeing as your on layer2 you'll need to reverse these things:
1. mac addresses
2. ipv4 addresses
3. ports

Then you will need to recalculate the header checksum. The header checksum is used to validate packets and if it's incorrect, the packets are dropped. 

Also, make sure when you write the packet that it's expecting a layer2 write.. maybe it only wants layer3 (ipv4) and up because it wants to add the mac addresses based on the internal arp table.

**Comment by gordonguilloryklx** [comment]
Okay got it. Thanks for you help. I'll let you know what I come up with. Thanks for the quick help!

**Comment by gordonguilloryklx** [comment]
@songgao  and @zveinn  I managed to swamp the source and destination (for this example) and recalculate the checksums, but windows will only show me data received, or sent not both.  Thinking you had me switch the src and dst because I was attempting to echo the data.

My goal is to use https://github.com/martin-ger/MQTT_VPN with this library and create a mqtt vpn tunnel to an iot device. I managed to get my code base up and sending data to my server, but any thing my other client (another pc for now) receives does not get wrote to the interface and vice versa.

My apologies, I am a bit new to packets, frames, etc. I am doing this project to learn them a bit better. If you have any other Idea let me know.
